### PR TITLE
PARQUET: Move StringColumnWriter dictionary to use string_t to avoid allocations

### PR DESCRIPTION
**Pull request by bleskes**
Note: normally copying `string_t` into an internal map requires cloning the data to avoid someone else releasing it. This is not needed in this case because the dictionary is only maintained while a single chunk is written and ParquetWriter's reference to it is sufficient to guarantee it doesn't change.

### CODE REVIEW

File: extension/parquet/column_writer.cpp [hunk 1]
```diff
@@ -1152,6 +1152,19 @@ class StringStatisticsState : public ColumnWriterStatistics {
...
+		return Hash(k);
+	}
+};
+
+struct StringEquality {
+	bool operator()(const string_t &a, const string_t &b) const {
+		return Equals::Operation(a, b);
+	}
+};
+typedef unordered_map<string_t, uint32_t, StringHash, StringEquality> string_dictionary_t;
```
Comment (Mytherin) [inline extension/parquet/column_writer.cpp:1166]: Could we move this to a separate header and unify it with the equivalent string dictionary in `dictionary_compression.cpp`?

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks great, one suggestion:

**Comment by bleskes** [comment]
Sure. I tried to move it to the string type header but that created a circular dependency so I let it go. I didn’t feel like it deserved a whole new header file on it’s own but happy to do so. Will ping when the pr is updated.

**Comment by bleskes** [comment]
@Mytherin I pushed an update. I see that node.js is failing but I'd appreciate some hint as to what I need to fix.. it's some missing import but duckdb compiles just fine.

**Comment by Mytherin** [comment]
The node.js library uses the parquet amalgamation and the DuckDB amalgamation to build, so any DuckDB includes (minus duckdb.hpp) need to be surrounded by `#ifndef DUCKDB_AMALGAMATION`.

**Comment by bleskes** [comment]
Thanks @Mytherin - I just wanted to double-check before making the change - other `.hpp` files in the common folder don't have `#ifndef DUCKDB_AMALGAMATION` in them, which may suggest I need a different solution for this one?

**Comment by Mytherin** [comment]
> Thanks @Mytherin - I just wanted to double-check before making the change - other `.hpp` files in the common folder don't have `#ifndef DUCKDB_AMALGAMATION` in them, which may suggest I need a different solution for this one?

All the duckdb files have this besides the include you added, no?

```cpp
#include "duckdb.hpp"
#ifndef DUCKDB_AMALGAMATION
#include "duckdb/common/common.hpp"
#include "duckdb/common/exception.hpp"
#include "duckdb/common/mutex.hpp"
#include "duckdb/common/serializer/buffered_file_writer.hpp"
#include "duckdb/common/types/chunk_collection.hpp"
#include "duckdb/common/types/date.hpp"
#include "duckdb/common/types/hugeint.hpp"
#include "duckdb/common/types/time.hpp"
#include "duckdb/common/types/timestamp.hpp"
#include "duckdb/common/serializer/buffered_serializer.hpp"
#include "duckdb/common/operator/comparison_operators.hpp"
#endif
 #include "snappy.h"
 #include "miniz_wrapper.hpp"
 #include "zstd.h"
 #include "duckdb/common/string_map_set.hpp"

```

**Comment by Mytherin** [comment]
Thanks! The CI failure is unrelated

**Pull request merged**